File: kotlin-node/karakum/plugins/convertCollections.js
import ts from "typescript";

export default function (node, context, render) {
    if (ts.isArrayTypeNode(node)) {
        return `js.array.ReadonlyArray<${render(node.elementType)}>`
    }

    if (
        ts.isIdentifier(node)
        && ts.isTypeReferenceNode(node.parent)
        && node.text === "Record"
    ) {
        return "js.objects.ReadonlyRecord"
    }

    if (
        ts.isIdentifier(node)
        && ts.isTypeReferenceNode(node.parent)
        && node.text === "Iterable"
    ) {
        return "js.iterable.JsIterable"
    }
    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertCollections.kt
package node.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.isArrayTypeNode
import typescript.isIdentifier
import typescript.isTypeReferenceNode

val convertCollections = createPlugin { node, _, render ->
    nullable {
        ensure(isArrayTypeNode(node))

        "js.array.ReadonlyArray<${render(node.elementType)}>"
    } ?: nullable {
        ensure(isIdentifier(node))
        ensure(isTypeReferenceNode(node.parent))
        ensure(node.text == "Record")

        "js.objects.ReadonlyRecord"
    } ?: nullable {
        ensure(isIdentifier(node))
        ensure(isTypeReferenceNode(node.parent))
        ensure(node.text == "Iterable")

        "js.iterable.JsIterable"
    }
}
